src/dune_rules/lib_rules.ml

```diff
@@ -570,86 +571,112 @@ let library_rules
   let* requires_compile = Compilation_context.requires_compile cctx in
   let ocaml = Compilation_context.ocaml cctx in
   let stdlib_dir = ocaml.lib_config.stdlib_dir in
-  let top_sorted_modules =
-    let impl_only = Modules.impl_only modules in
-    Dep_graph.top_closed_implementations
-      (Compilation_context.dep_graphs cctx).impl
-      impl_only
-  in
-  let* () =
-    Memo.Option.iter vimpl ~f:(Virtual_rules.setup_copy_rules_for_impl ~sctx ~dir)
-  in
-  let* () = Check_rules.add_cycle_check sctx ~dir top_sorted_modules in
-  let* () = gen_wrapped_compat_modules lib cctx
-  and* () = Module_compilation.build_all cctx
-  and* expander = Super_context.expander sctx ~dir
-  and* lib_info =
+  let* lib_info =
     let lib_config = ocaml.lib_config in
     let info = Library.to_lib_info lib ~dir ~lib_config in
     let mode = Lib_mode.Map.Set.for_merlin (Lib_info.modes info) in
     let+ () = Check_rules.add_obj_dir sctx ~obj_dir mode in
     info
   in
-  let+ () =
-    Memo.when_
-      (not (Library.is_virtual lib))
-      (fun () -> setup_build_archives lib ~lib_info ~top_sorted_modules ~cctx ~expander)
-  and+ () =
-    let vlib_stubs_o_files = Vimpl.vlib_stubs_o_files vimpl in
-    Memo.when_
-      (Library.has_foreign lib || List.is_non_empty vlib_stubs_o_files)
-      (fun () ->
-        build_stubs
-          lib
-          ~cctx
-          ~dir
-          ~expander
-          ~requires:requires_compile
-          ~dir_contents
-          ~vlib_stubs_o_files)
-  and+ () = Odoc.setup_library_odoc_rules cctx local_lib
-  and+ () =
-    Sub_system.gen_rules
-      { super_context = sctx; dir; stanza = lib; scope; source_modules; compile_info }
+  let* enabled_if = Lib_info.enabled lib_info in
+  let enabled =
+    match enabled_if with
+    | Disabled_because_of_enabled_if -> false
+    | Normal | Optional -> true
   in
-  ( cctx
-  , Merlin.make
-      ~requires:requires_compile
-      ~stdlib_dir
-      ~flags
-      ~modules
-      ~source_dirs:Path.Source.Set.empty
-      ~preprocess:(Preprocess.Per_module.without_instrumentation lib.buildable.preprocess)
-      ~libname:(Some (snd lib.name))
-      ~obj_dir
-      ~dialects:(Dune_project.dialects (Scope.project scope))
-      ~ident:(Lib.Compile.merlin_ident compile_info)
-      ~modes:(`Lib (Lib_info.modes lib_info)) )
+  match enabled with
```

Comment:
This check might look redundant, but if we don't conditionally return merlin config, the `eif-library-name-collision-same-folder` test fails with multiple rules for cmxs:

```
Error: Multiple rules generated for _build/alt-context/foo.cmxs:
- dune:4
- dune:1
Error: Multiple rules generated for _build/default/foo.cmxs:
- dune:4
- dune:1
```